library/std/src/net/udp.rs

```diff
@@ -643,12 +650,17 @@ impl UdpSocket {
     /// function of a UDP socket is not a useful thing to do: The OS will be
     /// unable to determine whether something is listening on the remote
     /// address without the application sending data.
+    ///
+    /// Linux: If your first `connect` is to a localhost address, subsequent
+    /// `connect`s to non-localhost addresses might fail.
```

Comment:
As a general rule, we don't tend to document platform-specific quirks because otherwise we wind up repeating all of e.g. [learn.microsoft.com](https://learn.microsoft.com)'s content. That doesn't mean we never provide hints on how to use a function, however.

I think we can accept this if you have a more general way of expressing it. Especially if there's an underlying principle at work that causes Linux to have this behavior and that will also cause any random OS to probably also do the same.

However, consulting the Windows documentation leads me to wonder: are you sure it was the _connect_ call that had this result, and not [some other function call performing an implicit bind](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-connect#remarks)?

Reply:
I only tested on linux. My knowledge doesn't go very deep in this topic, so finding an underlying principle would be very hard for me. As noted, I failed quite hard to find anyone else who had that problem before ([one thread](https://www.mail-archive.com/netdev@vger.kernel.org/msg159519.html)), so I can't read up easily on the topic. I already tried to navigate through the appropriate man pages but they didn't seem to include this error case.
> However, consulting the Windows documentation leads me to wonder: are you sure it was the connect call that had this result, and not [some other function call performing an implicit bind](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-connect#remarks)?

The documentation you linked mentions that `send_to` might perform an implicit bind, however I don't call that function. Also, I don't know to which extent the windows documentation aligns to linux behavior. The theory that an implicity bind happens somewhere does sound like a good lead.

```rs
fn main() {
    let socket = std::net::UdpSocket::bind("0.0.0.0:0").unwrap();
    socket.connect("192.168.2.100:3001").unwrap();
    socket.connect("127.0.0.1:1235").unwrap();
    socket.connect("192.168.2.100:3001").unwrap();
    // Success
}
fn main() {
    let socket = std::net::UdpSocket::bind("0.0.0.0:0").unwrap();
    // socket.connect("192.168.2.100:3001").unwrap();
    socket.connect("127.0.0.1:1235").unwrap();
    socket.connect("192.168.2.100:3001").unwrap(); // Error 22: InvalidInput, message: "Invalid argument"
}
fn main() {
    let socket = std::net::UdpSocket::bind("192.168.2.123:0").unwrap();
    // socket.connect("192.168.2.100:3001").unwrap();
    socket.connect("127.0.0.1:1235").unwrap();
    socket.connect("192.168.2.100:3001").unwrap();
    // Success
}
```
I could guess and say that maybe the kernel, after seeing the undefined address `0.0.0.0` binds to the network of the first `bind` address? However I honestly am not attached enough to this issue to start more extensive testing, networking isn't my thing sadly. Since this failure did cost me some time when trying to use this module, I do think that this should be documented somehow.

Reply:
You don't _need_ to do more extensive testing, really, as I'd be happy to see us document a more general remark here. It's already a hedged "might", after all.